File: backend/src/main/java/cn/etarch/mao/harness/delegate/SubagentExecutionRecoveryService.java
package cn.etarch.mao.harness.delegate;

import cn.etarch.mao.config.DelegateConfig;
import cn.etarch.mao.harness.core.AgentExecutionContext;
import cn.etarch.mao.harness.core.AgentLoop;
import cn.etarch.mao.harness.delegate.entity.SubagentExecution;
import cn.etarch.mao.harness.delegate.mapper.SubagentExecutionMapper;
import cn.etarch.mao.harness.local.LocalToolSessionRegistry;
import cn.etarch.mao.harness.tool.impl.DelegateTool;
import cn.etarch.mao.session.entity.Message;
import cn.etarch.mao.session.entity.Session;
import cn.etarch.mao.session.entity.SessionCompaction;
import cn.etarch.mao.session.mapper.MessageMapper;
import cn.etarch.mao.session.mapper.SessionMapper;
import cn.etarch.mao.session.service.SessionCompactionService;
import cn.etarch.mao.session.service.SessionService;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

@Slf4j
@Service
@RequiredArgsConstructor
public class SubagentExecutionRecoveryService {

    private final SubagentExecutionMapper executionMapper;
    private final SessionMapper sessionMapper;
    private final MessageMapper messageMapper;
    private final SessionService sessionService;
    private final SessionCompactionService compactionService;
    private final AgentDefinitionRegistry definitionRegistry;
    private final DelegateTool delegateTool;
    private final AgentLoop agentLoop;
    private final SubAgentVisibilityService visibilityService;
    private final LocalToolSessionRegistry localRegistry;
    private final DelegateConfig delegateConfig;
    private final ObjectMapper objectMapper;

    public void recover(Long executionId) {
        SubagentExecution execution = executionMapper.selectById(executionId);
        if (execution == null || isTerminal(execution.getStatus())) return;
        if (executionMapper.claimForRecovery(executionId) == 0) return;
        execution = executionMapper.selectById(executionId);
        Session parent = sessionMapper.selectById(execution.getParentSessionId());
        Session child = sessionMapper.selectById(execution.getChildSessionId());
        if (parent == null || child == null) {
            fail(execution, child, "子代理恢复失败：父会话或子会话不存在");
            return;
        }
        if (isTerminal(parent.getPhase())) {
            cancel(execution, child, "子代理已随父会话取消");
            return;
        }

        log.info("subagent_recovery_start executionId={} parent={} child={} invocation={}",
                executionId, parent.getId(), child.getId(), execution.getInvocationType());
        long started = System.currentTimeMillis();
        AtomicBoolean cancelFlag = agentLoop.registerCancelFlag(child.getId());
        boolean local = "LOCAL".equalsIgnoreCase(child.getExecutionMode());
        try {
            Message existingFinal = findExistingFinal(execution);
            if (existingFinal != null) {
                completeFromExistingFinal(execution, child, existingFinal);
                return;
            }

            SessionCompaction compaction = compactionService.loadValidated(child.getId());
            sessionService.cleanupIncompleteTailAfterId(child.getId(), compactionService.boundaryOf(compaction));
            sessionService.updatePhase(child.getId(), "RESUMING");

            AgentDefinition definition = definitionRegistry.getDefinition(execution.getAgentType());
            if (definition == null) throw new IllegalStateException("未知的子代理类型: " + execution.getAgentType());
            AgentExecutionContext context = delegateTool.buildSubContext(child, definition);
            context.setCancelFlag(cancelFlag);

            long deadline = startedDeadline(execution.getStartedAt(), delegateConfig.getTimeoutSeconds());
            if (deadline <= System.currentTimeMillis()) {
                throw new IllegalStateException("子代理恢复失败：原委派执行已超时");
            }
            if (local) {
                localRegistry.setUserForSession(child.getId(), parent.getUserId());
                waitForLocal(child.getId(), deadline, cancelFlag);
            }

            long remainingSeconds = Math.max(1, (deadline - System.currentTimeMillis()) / 1000);
            SubAgentVisibilityService.VisibleRunResult run = visibilityService.executeVisibleWithTimeout(
                    child, context, false, cancelFlag, remainingSeconds, delegateConfig.getCancelGraceSeconds());
            SubAgentResultCollector collector = run.getCollector();
            if (cancelFlag.get()) {
                cancel(execution, child, "子代理已随父会话取消");
            } else if (collector.getError() != null) {
                fail(execution, child, "子代理恢复失败：" + collector.getError().getMessage());
            } else {
                String result = collector.getResult().isBlank() ? "(子代理未产生文本输出)" : collector.getResult();
                Message finalMessage = latestAssistantAfterStart(execution);
                if (finalMessage == null || collector.getResult().isBlank()) {
                    finalMessage = sessionService.saveMessage(child.getId(), "ASSISTANT", result,
                            collector.getThinkingContent(), null, null,
                            collector.getTotalUsage() != null ? collector.getTotalUsage().getTotalTokens() : 0,
                            context.getModelConfig() != null ? context.getModelConfig().getId() : null);
                }
                terminal(execution, "COMPLETED", result, context.getCurrentRound(), collector, finalMessage.getId());
                visibilityService.finishSubagent(child.getId(), parent.getUserId(), "COMPLETED", run.getExecutionId());
            }
        } catch (Exception e) {
            fail(execution, child, "子代理恢复失败：" + safeMessage(e));
        } finally {
            agentLoop.removeCancelFlag(child.getId());
            if (local) localRegistry.removeSession(child.getId());
            SubagentExecution done = executionMapper.selectById(executionId);
            log.info("subagent_recovery_complete executionId={} status={} durationMs={}", executionId,
                    done != null ? done.getStatus() : "UNKNOWN", System.currentTimeMillis() - started);
        }
    }

    private void waitForLocal(Long childId, long deadline, AtomicBoolean cancelFlag) throws InterruptedException {
        while (!localRegistry.isConnected(childId) && !cancelFlag.get() && System.currentTimeMillis() < deadline) {
            TimeUnit.MILLISECONDS.sleep(Math.min(250, Math.max(1, deadline - System.currentTimeMillis())));
        }
        if (cancelFlag.get()) {
            throw new IllegalStateException("LOCAL 客户端等待已取消");
        }
        if (!localRegistry.isConnected(childId)) {
            throw new IllegalStateException("子代理恢复失败：LOCAL 客户端未在恢复超时内连接");
        }
    }

    private long startedDeadline(LocalDateTime startedAt, int timeoutSeconds) {
        long started = startedAt != null
                ? startedAt.atZone(java.time.ZoneId.systemDefault()).toInstant().toEpochMilli()
                : System.currentTimeMillis();
        return started + timeoutSeconds * 1000L;
    }

    private Message findExistingFinal(SubagentExecution execution) {
        if (execution.getFinalMessageId() != null) {
            Message message = messageMapper.selectById(execution.getFinalMessageId());
            if (message != null && "ASSISTANT".equals(message.getRole())) return message;
        }
        return latestAssistantAfterStart(execution);
    }

    private Message latestAssistantAfterStart(SubagentExecution execution) {
        if (execution.getExecutionStartMessageId() == null) return null;
        return messageMapper.selectOne(new LambdaQueryWrapper<Message>()
                .eq(Message::getSessionId, execution.getChildSessionId())
                .eq(Message::getRole, "ASSISTANT")
                .gt(Message::getId, execution.getExecutionStartMessageId())
                .isNull(Message::getToolCalls)
                .orderByDesc(Message::getId).last("LIMIT 1"));
    }

    private void completeFromExistingFinal(SubagentExecution execution, Session child, Message message) {
        String status = terminalStatus(message);
        terminal(execution, status, message.getContent(), execution.getTotalRounds(), null, message.getId());
        sessionService.updatePhase(child.getId(), status);
    }

    private String terminalStatus(Message message) {
        if (message.getMetadata() == null || message.getMetadata().isBlank()) return "COMPLETED";
        try {
            String status = objectMapper.readTree(message.getMetadata()).path("subagentTerminalStatus").asText();
            return "FAILED".equals(status) || "CANCELLED".equals(status) ? status : "COMPLETED";
        } catch (Exception ignored) {
            return "COMPLETED";
        }
    }

    private void fail(SubagentExecution execution, Session child, String result) {
        Message finalMessage = child != null ? sessionService.saveMessage(child.getId(), "ASSISTANT", result,
                null, null, null, 0, null,
                "{\"subagentTerminalStatus\":\"FAILED\"}") : null;
        terminal(execution, "FAILED", result, execution.getTotalRounds(), null,
                finalMessage != null ? finalMessage.getId() : null);
        if (child != null) visibilityService.finishSubagent(child.getId(), child.getUserId(), "FAILED", null);
    }

    private void cancel(SubagentExecution execution, Session child, String result) {
        terminal(execution, "CANCELLED", result, execution.getTotalRounds(), null, null);
        if (child != null) visibilityService.finishSubagent(child.getId(), child.getUserId(), "CANCELLED", null);
    }

    private void terminal(SubagentExecution execution, String status, String result, Integer rounds,
                          SubAgentResultCollector collector, Long finalMessageId) {
        execution.setStatus(status);
        execution.setResult(truncate(result));
        execution.setCompletedAt(LocalDateTime.now());
        execution.setFinalMessageId(finalMessageId);
        if (rounds != null) execution.setTotalRounds(rounds);
        if (collector != null) {
            execution.setTotalToolCalls(collector.getToolCallCount());
            if (collector.getTotalUsage() != null) {
                execution.setTotalPromptTokens(collector.getTotalUsage().getPromptTokens());
                execution.setTotalCompletionTokens(collector.getTotalUsage().getCompletionTokens());
            }
        }
        executionMapper.updateById(execution);
    }

    private boolean isTerminal(String status) {
        return "COMPLETED".equals(status) || "FAILED".equals(status) || "CANCELLED".equals(status);
    }

    private String truncate(String value) {
        return value != null && value.length() > 65000 ? value.substring(0, 65000) : value;
    }

    private String safeMessage(Exception e) {
        return e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName();
    }
}

File: backend-ts/src/harness/delegate/subagent-execution-recovery.service.ts
import type { SubagentExecution } from '../../session/types.js';
import type { AgentLoop } from '../core/agent-loop.js';
import type { Message, SessionCompactionService, SessionMapper, SessionService } from '../deps.js';
import type { LocalToolSessionRegistry } from '../local/local-tool-session-registry.js';
import { harnessLog } from '../log.js';
import type { DelegateTool } from '../tool/impl/delegate-tool.js';
import type { AgentDefinitionRegistry } from './agent-definition-registry.js';
import type { SubagentExecutionMapper } from './subagent-execution.mapper.js';
import type { SubAgentVisibilityService } from './subagent-visibility-service.js';

export class SubagentExecutionRecoveryService {
  constructor(
    private readonly executionMapper: SubagentExecutionMapper,
    private readonly sessionMapper: SessionMapper,
    private readonly sessionService: SessionService,
    private readonly compactionService: SessionCompactionService,
    private readonly definitionRegistry: AgentDefinitionRegistry,
    private readonly delegateTool: DelegateTool,
    private readonly agentLoop: AgentLoop,
    private readonly visibilityService: SubAgentVisibilityService,
    private readonly localRegistry: LocalToolSessionRegistry,
    private readonly timeoutSeconds = 3600,
    private readonly cancelGraceSeconds = 30,
  ) {}

  async recover(execution: SubagentExecution): Promise<void> {
    if (execution.id == null || execution.childSessionId == null) return;
    if (!['RUNNING', 'RECOVERING'].includes(execution.status ?? '')) return;
    const parent = execution.parentSessionId != null
      ? await this.sessionMapper.selectById(execution.parentSessionId) : null;
    const child = await this.sessionMapper.selectById(execution.childSessionId);
    if (!parent || isTerminal(parent.phase)) {
      await this.executionMapper.updateTerminal(execution.id, {
        status: 'CANCELLED', result: '子代理已随父会话取消', completedAt: nowSql(),
      });
      return;
    }
    if (!child) {
      await this.fail(execution, '子代理恢复失败：原子会话不存在');
      return;
    }
    const definition = this.definitionRegistry.getDefinition(execution.agentType ?? '');
    if (!definition) {
      await this.fail(execution, `子代理恢复失败：未知的子代理类型 ${execution.agentType ?? ''}`);
      return;
    }
    await this.executionMapper.claimRecovering(execution.id);
    harnessLog('info', `subagent_recovery_start executionId=${execution.id} parent=${execution.parentSessionId} child=${child.id} invocation=${execution.invocationType ?? 'legacy'}`);
    try {
      const existingFinal = await this.findPersistedFinal(execution);
      if (existingFinal) {
        const status = terminalStatus(existingFinal.metadata);
        await this.executionMapper.updateTerminal(execution.id, {
          status, result: existingFinal.content ?? '(子代理未产生文本输出)',
          finalMessageId: existingFinal.id, completedAt: nowSql(),
        });
        await this.sessionService.updatePhase(child.id!, status);
        return;
      }
      const compaction = await this.compactionService.loadValidated(child.id!);
      const boundary = this.compactionService.boundaryOf(compaction);
      await this.sessionService.cleanupIncompleteTailAfterId?.(child.id!, boundary);
      await this.sessionService.updatePhase(child.id!, 'RESUMING');
      const context = await this.delegateTool.buildSubContext(child, definition);
      const cancel = this.agentLoop.registerCancelFlag(child.id!);
      context.cancelFlag = cancel;
      const deadline = startedDeadline(execution.startedAt, this.timeoutSeconds);
      if (deadline <= Date.now()) throw new Error('子代理恢复失败：原委派执行已超时');
      if (child.executionMode?.toUpperCase() === 'LOCAL') {
        if (parent.userId != null) {
          this.localRegistry.setUserForSession(parent.id!, parent.userId);
          this.localRegistry.setUserForSession(child.id!, parent.userId);
        }
        const connected = await this.waitForLocal(child.id!, deadline, cancel);
        if (!connected) throw new Error('子代理恢复失败：LOCAL 客户端未在恢复超时内连接');
      }
      const remainingSeconds = Math.max(1, Math.floor((deadline - Date.now()) / 1000));
      const run = await this.visibilityService.executeVisibleWithTimeout(
        child, context, false, cancel, remainingSeconds, this.cancelGraceSeconds,
      );
      const collector = run.collector;
      const cancelled = cancel.get();
      let status = cancelled ? 'CANCELLED' : collector.error ? 'FAILED' : 'COMPLETED';
      let result = collector.getResult();
      if (!result) {
        result = status === 'COMPLETED'
          ? '(子代理未产生文本输出)'
          : `子代理恢复失败：${collector.error instanceof Error ? collector.error.message : '子代理执行异常'}`;
        await this.sessionService.saveMessage(
          child.id!, 'ASSISTANT', result, collector.getThinkingContent(), null, null, 0,
          context.modelConfig?.id ?? null,
          status === 'FAILED' ? JSON.stringify({ subagentTerminalStatus: 'FAILED' }) : null,
        );
      }
      const final = await this.findLastAssistant(child.id!, execution.executionStartMessageId ?? 0);
      await this.executionMapper.updateTerminal(execution.id, {
        status, result, finalMessageId: final?.id ?? null,
        totalRounds: context.currentRound,
        totalPromptTokens: collector.totalUsage?.promptTokens ?? 0,
        totalCompletionTokens: collector.totalUsage?.completionTokens ?? 0,
        totalToolCalls: collector.toolCallCount,
        completedAt: nowSql(),
      });
      await this.visibilityService.finishSubagent(child.id!, parent.userId, status, run.executionId);
    } catch (error) {
      await this.fail(execution, `子代理恢复失败：${(error as Error).message}`);
      await this.visibilityService.finishSubagent(child.id!, parent.userId, 'FAILED', '');
    } finally {
      this.agentLoop.removeCancelFlag(child.id!);
      this.localRegistry.removeSession?.(child.id!);
      harnessLog('info', `subagent_recovery_complete executionId=${execution.id}`);
    }
  }

  private async findPersistedFinal(execution: SubagentExecution): Promise<Message | null> {
    if (execution.finalMessageId != null) {
      const messages = await this.sessionService.getMessagesAfterId(execution.childSessionId!, execution.finalMessageId - 1);
      return messages.find((message) => message.id === execution.finalMessageId && message.role === 'ASSISTANT') ?? null;
    }
    if (execution.executionStartMessageId == null) return null;
    const messages = await this.sessionService.getMessagesAfterId(execution.childSessionId!, execution.executionStartMessageId);
    return [...messages].reverse().find((message) => message.role === 'ASSISTANT' && !message.toolCalls) ?? null;
  }

  private async findLastAssistant(childSessionId: number, startMessageId: number): Promise<Message | null> {
    const messages = await this.sessionService.getMessagesAfterId(childSessionId, startMessageId);
    return [...messages].reverse().find((message) => message.role === 'ASSISTANT' && !message.toolCalls) ?? null;
  }

  private async waitForLocal(
    childSessionId: number,
    deadline: number,
    cancel: { get(): boolean },
  ): Promise<boolean> {
    while (!cancel.get() && Date.now() < deadline) {
      if (await this.localRegistry.isConnected(childSessionId)) return true;
      await sleep(Math.min(1000, Math.max(1, deadline - Date.now())));
    }
    return false;
  }

  private async fail(execution: SubagentExecution, message: string): Promise<void> {
    if (execution.id == null) return;
    const childId = execution.childSessionId;
    if (childId != null) {
      const existing = await this.findLastAssistant(childId, execution.executionStartMessageId ?? 0);
      if (!existing) {
        await this.sessionService.saveMessage(
          childId, 'ASSISTANT', message, null, null, null, 0, null,
          JSON.stringify({ subagentTerminalStatus: 'FAILED' }),
        );
      }
      await this.sessionService.updatePhase(childId, 'FAILED');
    }
    await this.executionMapper.updateTerminal(execution.id, {
      status: 'FAILED', result: message, completedAt: nowSql(),
    });
  }
}

function terminalStatus(metadata: string | null | undefined): 'COMPLETED' | 'FAILED' | 'CANCELLED' {
  if (!metadata) return 'COMPLETED';
  try {
    const status = (JSON.parse(metadata) as { subagentTerminalStatus?: string }).subagentTerminalStatus;
    return status === 'FAILED' || status === 'CANCELLED' ? status : 'COMPLETED';
  } catch {
    return 'COMPLETED';
  }
}

function startedDeadline(startedAt: string | null | undefined, timeoutSeconds: number): number {
  const started = startedAt ? new Date(startedAt.replace(' ', 'T') + 'Z').getTime() : Date.now();
  return (Number.isFinite(started) ? started : Date.now()) + timeoutSeconds * 1000;
}

function isTerminal(phase: string | null | undefined): boolean {
  return phase === 'COMPLETED' || phase === 'FAILED' || phase === 'CANCELLED';
}

function nowSql(): string {
  return new Date().toISOString().replace('T', ' ').slice(0, 19);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
